### DISCUSSION TITLE
[1.1.454] ControlNet union SDXL model

### DISCUSSION THREAD

**Discussion by huchenlei** [Mikubill/sd-webui-controlnet#2989] [General]
ControlNet Union is a new type of ControlNet that can handle multiple input conditions
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/15da4e01-a8cb-4e18-8eb6-9f74d4cfcec5)

- https://huggingface.co/xinsir/controlnet-union-sdxl-1.0
- https://github.com/xinsir6/ControlNetPlus

# How to use
## Download model
Model can be downloaded here: https://huggingface.co/xinsir/controlnet-union-sdxl-1.0/tree/main. After download, it is recommended to rename the file as `controlnet++_union_sdxl` so that the control type filter can correctly handle it.

## Openpose
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/2f08888d-0adc-4677-ae41-678181674a97)
![00021-1413879340](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/e0426c7b-9a7e-473c-8e61-9b777ed5d753)

Do not use key map with face and hand as the union controlnet does not seem to be trained with hand/face annotation
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/702a071c-dec2-4dd9-bf96-87f9b9f8375b)
![00020-1413879340](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/b6b7188d-890c-4e64-a336-9f790e5bf10d)

## Canny
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/625df0a1-9580-48eb-b7bc-f65c052c6804)
![00017-1413879340](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/7f4b7663-3eb6-4bdd-b3f0-49ab8972cf73)

## Depth
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/7c98e394-60b8-4718-b7b8-7bcd54e04dab)
![00018-1413879340](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/cba1f5d7-5a23-4968-a1dc-02ac2707a538)

## SoftEdge
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/c8e47ce5-2673-4029-83d8-c0b72d351609)
![00019-1413879340](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/08198e17-7fea-4026-ab00-7ce11fe0f87a)

## Normal
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/dff25b2b-9755-48ae-acb7-9333bd692fcb)
![00022-1413879340](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/99d9af38-15b0-4bed-9ae5-a6f34bf88553)

## Segmentation
Note: `ofade20k` preprocessor does not seem to work
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/569dcf7d-7ade-4174-b293-c5b79bd1004d)
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/2e5ce5e1-a8b9-4460-8357-2eae5182e42d)

# Multi-ControlNet
Currently the multi-controlnet is not working in the optimal way described in the original paper, but you can still try use it, as it can help you save VRAM by avoid loading another controlnet for different type of control.

## Unit1 setup
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/916d792b-3399-438d-b2e3-16dcb68cf568)

## Unit 2 setup
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/9fcb1cc6-10e0-446a-919f-ff1887297f1c)

## Generation result
![00026-1413879340](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/f10d0e77-e37f-4ad3-8d89-10f2434ceb15)
 
# Appendix
Input images
![image (1)](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/8c29b500-73a0-4b85-8534-efa3297fccb0)
![input](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/75a5f008-dd44-45ca-8421-367ae8bee70e)

Mask
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/83ae4e69-0625-4514-beb7-080742d79d66)

Generation parameters (Parameters including seed are fixed for all examples)
```
1girl, sitting on throne,, absurdres, highres, (masterpiece:1.2), (best quality, highest quality),
Negative prompt: (lowres, low quality, worst quality:1.2), (text:1.2), watermark, (frame:1.2), deformed, ugly, deformed eyes, blur, out of focus, blurry, deformed cat, deformed, photo, anthropomorphic cat, monochrome, photo, pet collar, gun, weapon, blue, 3d, drones, drone, buildings in background, green
Steps: 30, Sampler: DPM++ 2M, Schedule type: Karras, CFG scale: 7, Seed: 1413879340, Size: 1024x1024, Model hash: 5d973c1f8a, Model: animagine-xl-2.0, ControlNet 0: "Module: none, Model: controlnet++_union_sdxl [15e6ad5d], Weight: 1.0, Resize Mode: Crop and Resize, Processor Res: 512, Threshold A: 0.5, Threshold B: 0.5, Guidance Start: 0.0, Guidance End: 1.0, Pixel Perfect: False, Control Mode: Balanced", ControlNet 1: "Module: canny, Model: controlnet++_union_sdxl [15e6ad5d], Weight: 1.0, Resize Mode: Crop and Resize, Processor Res: 512, Threshold A: 100.0, Threshold B: 200.0, Guidance Start: 0.0, Guidance End: 1.0, Pixel Perfect: False, Control Mode: Balanced", Version: v1.9.4-249-g93c00b2a
```

**Comment by sdbds** [comment]
https://github.com/xinsir6/ControlNetPlus/issues/1
According to my issue, union used mobile sam random split instead of ADE20K

Considering the [controlnet_aux](https://github.com/huggingface/controlnet_aux) repository is now hosted by huggingface, and more new research papers will use the controlnet_aux package, I think we can talk to @Fannovel16 about unifying the preprocessor parts of the three projects to update controlnet_aux.

**Comment by Fannovel16** [comment]
@sdbds It's quite challenging as `comfyui_controlnet_aux` has a different DWPose [implementation](https://github.com/Fannovel16/comfyui_controlnet_aux/tree/main/src/controlnet_aux/dwpose) and embeds many [third party packages's fork](https://github.com/Fannovel16/comfyui_controlnet_aux/tree/main/src)

**Reply by sdbds** [reply]
Yes, I'm going to start experimenting with preprocessors that don't rely on third-party packages, and I'll just use what's in the repository for the rest of it.
Also controlnet_aux's DWpose implementation seems to be faster and more suitable for video frame conversion, in my previous tests.

**Comment by slashedstar** [comment]
CAn we have a update for the promax version with tile and inpainting? 👀

**Reply by ali0une** [reply]
Yes please @huchenlei !

i currently have this error with the [promax](https://huggingface.co/xinsir/controlnet-union-sdxl-1.0/tree/main) version :

```
2024-07-14 17:47:35,187 - ControlNet - INFO - Loading model: controlnet++_union_sdxl-promax-1.0 [9460e4db]
2024-07-14 17:47:35,196 - ControlNet - INFO - Loaded state_dict from [/whatever/stable-diffusion-webui/models/ControlNet/controlnet++_union_sdxl-promax-1.0.safetensors]
2024-07-14 17:47:35,201 - ControlNet - INFO - controlnet_sdxl_config
*** Error running process: /whatever/stable-diffusion-webui/extensions/sd-webui-controlnet/scripts/controlnet.py
    Traceback (most recent call last):
      File "/whatever/stable-diffusion-webui/modules/scripts.py", line 825, in process
        script.process(p, *script_args)
      File "/whatever/stable-diffusion-webui/extensions/sd-webui-controlnet/scripts/controlnet.py", line 1228, in process
        self.controlnet_hack(p)
      File "/whatever/stable-diffusion-webui/extensions/sd-webui-controlnet/scripts/controlnet.py", line 1213, in controlnet_hack
        self.controlnet_main_entry(p)
      File "/whatever/stable-diffusion-webui/extensions/sd-webui-controlnet/scripts/controlnet.py", line 919, in controlnet_main_entry
        model_net, control_model_type = Script.load_control_model(p, unet, unit.model)
      File "/whatever/stable-diffusion-webui/extensions/sd-webui-controlnet/scripts/controlnet.py", line 436, in load_control_model
        control_model = Script.build_control_model(p, unet, model)
      File "/whatever/stable-diffusion-webui/extensions/sd-webui-controlnet/scripts/controlnet.py", line 465, in build_control_model
        control_model = build_model_by_guess(state_dict, unet, model_path)
      File "/whatever/stable-diffusion-webui/extensions/sd-webui-controlnet/scripts/controlnet_model_guess.py", line 240, in build_model_by_guess
        network = PlugableControlModel(config, state_dict)
      File "/whatever/stable-diffusion-webui/extensions/sd-webui-controlnet/scripts/cldm.py", line 24, in __init__
        self.control_model.load_state_dict(state_dict, strict=False)
      File "/whatever/stable-diffusion-webui/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 2189, in load_state_dict
        raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
    RuntimeError: Error(s) in loading state_dict for ControlNet:
    	size mismatch for task_embedding: copying a param with shape torch.Size([8, 320]) from checkpoint, the shape in current model is torch.Size([6, 320]).
    	size mismatch for control_add_embedding.linear_1.weight: copying a param with shape torch.Size([1280, 2048]) from checkpoint, the shape in current model is torch.Size([1280, 1536]).
```

**Reply by huchenlei** [reply]
Tile supported, but inpaint still has issue. https://github.com/Mikubill/sd-webui-controlnet/pull/2998

Need to wait official inference script for inpaint from union controlnet author to see what is wrong for inpaint.

**Reply by TKoestlerx** [reply]
Edit: Outpaint is working with the preprocessor set to "none". 
Inpaint is not working correctly as it seems to draw information from the masked out area.

Edit2: You can do inpainting if you use the "Effective Region Mask" Feature to upload the mask instead of the default inpainting map.

![small](https://github.com/user-attachments/assets/8cbfee86-adec-44f1-8e95-a78c9f084ac1)
![big1](https://github.com/user-attachments/assets/2de41ed9-f8ce-4e65-8f6c-cd12286f3d9b)
![big2](https://github.com/user-attachments/assets/0ffa48e5-24c4-48bd-8ecd-5e3596ce41c6)

**Reply by Gavr728** [reply]
I tried to do outpainting with the preprocessor set to none, but surprisingly it also changed colors on the original picture while preserving all the other details in tact.

**Reply by sdbds** [reply]
> Tile supported, but inpaint still has issue. #2998
> 
> Need to wait official inference script for inpaint from union controlnet author to see what is wrong for inpaint.

They put on other place...
https://github.com/xinsir6/ControlNetPlus/blob/main/promax/controlnet_union_test_inpainting.py

**Reply by AviceProjects** [reply]
Any updates to fixing inpainting? Maybe with what sdbds mentioned?

**Comment by ahmed20247** [comment]
Does it work with Domfyui?
